MDN React Tutorials #6: Accessibility & Resources (#31139)
mdn/content: files/en-us/learn/tools_and_testing/client-side_javascript_frameworks/react_accessibility/index.md

But where did our focus indicator go?

When we switch between templates in our `<Todo />` component, we completely remove the elements from the old template and replace them with the elements from the new template. That means the element that we were focused on no longer exists, so there's no visual cue as to where the browser's focus is. This could confuse a wide variety of users — particularly users who rely on the keyboard, or users who use assistive technology.

To improve the experience for keyboard and assistive technology users, we should manage the browser's focus ourselves.

### Aside: a note on our focus indicator

If you click the "All", "Active", or "Completed" filter buttons with your mouse, you _won't_ see a visible focus indicator, but you will do if you move between them with the tab key on your keyboard. Don't worry — your code isn't broken!

> Our CSS file uses the [`:focus-visible`](/en-US/docs/Web/CSS/:focus-visible) pseudo-class to provide custom styling for the focus indicator, and the browser uses a set of internal rules to determine when to show it to the user. Generally, the browser _will_ show a focus indicator in response to keyboard input, and _might_ show it in response to mouse input. `<button>` elements generally _don't_ show a focus indicator in response to mouse input, while `<input>` elements _do_.

Author: I would strike "generally" from "button elements generally don't show a focus indicator". I'm 95 percent sure that the heuristic of focus-visible is tag-based,and button elements will never show a visible indicator.

Reviewer: OK, done.

But where did our focus indicator go?

When we switch between templates in our `<Todo />` component, we completely remove the elements from the old template and replace them with the elements from the new template. That means the element that we were focused on no longer exists, so there's no visual cue as to where the browser's focus is. This could confuse a wide variety of users — particularly users who rely on the keyboard, or users who use assistive technology.

To improve the experience for keyboard and assistive technology users, we should manage the browser's focus ourselves.

### Aside: a note on our focus indicator

> If you click the "All", "Active", or "Completed" filter buttons with your mouse, you _won't_ see a visible focus indicator. Your code isn't broken! Our CSS uses the [`:focus-visible`](/en-US/docs/Web/CSS/:focus-visible) pseudo-class, which allows the browser to decide when to show a visual focus indicator to the users. Most often, this means an indicator will only appear when the user is navigating with the keyboard, but there are some exceptions: you'll notice that the indicator still shows up when you click the "What needs to be done?" text input at the top of the page.

Reviewer: @mxmason I think this info is useful, and it is also a good place for it. However, I do have a couple of issues to raise, which I think could warrant a bit of attention:

- `:focus-visible` doesn't allow the browser to decide when to show a visual focus indicator to the users — the browser does that anyway, using a set of internal heuristics. If you remove your `:focus-visible` style from the stylesheet you'll still get a focus indicator, it'll just be the browser default. What it does do is allow you to customize the focus indicator styling, without having to specify that you want it to appear on hover or some other state.
- It might be nice to say a little bit more about why the focus indicator doesn't appear when you click buttons, but it does when you focus an input field. Probably something along the lines of it not being needed on a button for mouse users as you just click and forget and its position isn't important like it is for a keyboard user. However, with an input it is important to help the user remember that they're on an input and still have a value to input...?

> If you click the "All", "Active", or "Completed" filter buttons with your mouse, you _won't_ see a visible focus indicator. Your code isn't broken! Our CSS uses the [`:focus-visible`](/en-US/docs/Web/CSS/:focus-visible) pseudo-class, which allows the browser to decide when to show a visual focus indicator to the users. Most often, this means an indicator will only appear when the user is navigating with the keyboard, but there are some exceptions: you'll notice that the indicator still shows up when you click the "What needs to be done?" text input at the top of the page.

Author: @chrisdavidmills Okay, fair enough. The reason I said that the browser decides, though, is that that's what `:focus-visible` is – a set of heuristics internal to the user agent. You see the same behavior without our CSS because user agents use `:focus-visible` as a default now, instead of our old friend `:focus`

As for explaining why, I agree with your reasoning as to why browser devs did what they did, but there's no real documentation to prove that, so I shied away from writing something that could be seen as speculative. Right now, :focus-visible is really just "this is a browser suggestion"

> If you click the "All", "Active", or "Completed" filter buttons with your mouse, you _won't_ see a visible focus indicator. Your code isn't broken! Our CSS uses the [`:focus-visible`](/en-US/docs/Web/CSS/:focus-visible) pseudo-class, which allows the browser to decide when to show a visual focus indicator to the users. Most often, this means an indicator will only appear when the user is navigating with the keyboard, but there are some exceptions: you'll notice that the indicator still shows up when you click the "What needs to be done?" text input at the top of the page.

Reviewer: You're not wrong on any of this, but the text does talk specifically about the `:focus-visible` pseudo class, and not the wider concept. I think you can leave further explanation of the behavior itself (2nd bullet), but I think you should update the text a little to make sure it isn't perceived as inaccurate. Maybe a combination of what we both said - say that in general browser heuristics control focus visibility, and then go on to say that it's visual output can be customized using the `:focus-visible` pseudo?

> If you click the "All", "Active", or "Completed" filter buttons with your mouse, you _won't_ see a visible focus indicator. Your code isn't broken! Our CSS uses the [`:focus-visible`](/en-US/docs/Web/CSS/:focus-visible) pseudo-class, which allows the browser to decide when to show a visual focus indicator to the users. Most often, this means an indicator will only appear when the user is navigating with the keyboard, but there are some exceptions: you'll notice that the indicator still shows up when you click the "What needs to be done?" text input at the top of the page.

Reviewer: Much better. I gave the section a bit of editorial tweaking. Have a look and let me know if you are OK with it.

> ### Aside: a note on our focus indicator

Author: 🔍  I added some heft to this section.

Reviewer: Cool, nice one. Just check my edits to see if you are OK with them.

Our CSS file uses the [`:focus-visible`](/en-US/docs/Web/CSS/:focus-visible) pseudo-class to define the focus indicator we see, and this pseudo-class uses a browser-defined set of rules to determine whether or not to show the focus indicator to the user. Generally, `:focus-visible` _will_ show a focus indicator in response to keyboard input, and _might_ show in response to mouse input. `<button>` elements are an example of an element that does not show a focus indicator in response to mouse input, while `<input>` elements _do_ show a focus indicator in response to mouse input.

The behavior of `:focus-visible` is more selective than the older [`:focus`](/en-US/docs/Web/CSS/:focus) pseudo-class, with which you might be more familiar. `:focus` shows a focus indicator in many more situations, and you can use it instead of or in combination with `:focus-visible` if you prefer.

## Focusing between templates

When a user toggles a `<Todo />` template from viewing to editing, we should focus on the `<input>` used to rename it; when they toggle back from editing to viewing, we should move focus back to the "Edit" button.

### Targeting our elements

> Up to this point, we've been writing JSX and letting React build the DOM from them. Most of the time, we don't need to target specific elements in the DOM because we can use React's state and props to control what gets rendered. To manage focus, however, we _do_ need to be able to target specific elements.

Author: 🔍 This section now moves into the `useEffect` section that follows a bit more nicely. By taking a bit more time to explain how `useRef` works,  my hope is that onboarding to `useEffect` will be less steep. I also added a specific explanation for `usePrevious` farther on.

</button>
```

> Doing this will populate our `editFieldRef` and `editButtonRef` with references to the DOM elements they're attached to, but _only_ after React has rendered the component. Test that out for yourself: add the following line somewhere in the body of your `Todo()` function:

Reviewer: ```suggestion
Doing this will populate our `editFieldRef` and `editButtonRef` with references to the DOM elements they're attached to, but _only_ after React has rendered the component. Test that out for yourself: add the following line somewhere in the body of your `Todo()` function, below where `editButtonRef` is initialized:
```

```jsx
console.log(editButtonRef.current);
```

You'll see that the value of `editButtonRef.current` is `null` when the component first renders, but if you click an "Edit" button, it will log the `<input>` element to the console. This is because the ref is populated only after the component renders, and clicking the "Edit" button causes the component to re-render. Be sure to delete this log before moving on.

> **Note:** Your logs will appear 6 times because we have 3 instances of `<Todo />` in our app and React renders our components twice in development.

We're getting closer! To take advantage of our newly referenced elements, we need to use another React hook: `useEffect()`.

### Implementing `useEffect()`

> [`useEffect()`](https://react.dev/reference/react/useEffect) is so named because it runs any side-effects that we'd like to add to the render process, and which we can't run inside the main function body. `useEffect()` runs right after a component renders, and this means that meaning the DOM elements we referenced in the previous section will be available for us to use.

Reviewer: ```suggestion
[`useEffect()`](https://react.dev/reference/react/useEffect) is so named because it runs any side-effects that we'd like to add to the render process, which can't be run inside the main function body. `useEffect()` runs right after a component renders, meaning the DOM elements we referenced in the previous section will be available for us to use.
```

Change the import statement of `Todo.jsx` again to add `useEffect`:

```jsx
import { useEffect, useRef, useState } from "react";
```

> `useEffect()` takes a function as an argument; this function is executed _after_ the component renders. To illustrate this, put the following `useEffect()` call just above the `return` statement in the body of `Todo()`, and pass a function into it that logs the words "side effect" to your console:

Reviewer: ```suggestion
`useEffect()` takes a function as an argument; this function is executed _after_ the component renders. To demonstrate this, put the following `useEffect()` call just above the `return` statement in the body of `Todo()`, and pass a function into it that logs the words "side effect" to your console:
```

Reviewer: Just so we aren't using "illustrate" twice in quick succession.

```jsx
if (wasNotEditingBefore && isEditingNow) {
  focusOnEditField();
} else if (wasEditingBefore && isNotEditingNow) {
  focusOnEditButton();
}
```

The React team had discussed [ways to get a component's previous state](https://reactjs.org/docs/hooks-faq.html#how-to-get-the-previous-props-or-state), and has provided an example custom hook we can use for the job.

> Paste the following code near the top of `Todo.jsx`, above your `Todo()` function.

Author: Yep, I argued with myself about writing this out because it's a little complicated, but it's probably worthwhile. Will do.

> Paste the following code near the top of `Todo.jsx`, above your `Todo()` function.

Reviewer: Cool, this new text is perfect.

}, [isEditing]);
```

This kind of works. If you use your keyboard to trigger the "Edit" button (remember: <kbd>Tab</kbd> to it and press <kbd>Enter</kbd>), you'll see that your focus moves between Edit `<input>` and "Edit" button as you start and end an edit. However, you may have noticed a new problem — the "Edit" button in the final `<Todo />` component is focused immediately on page load, before we even interact with the app!

> Our `useEffect()` hook is behaving exactly as we designed it: it runs as soon as the component renders, sees that `isEditing` is `false`, and focuses the "Edit" button. Because there are three instances of `<Todo />`, the last `<Todo />` that renders moves focus to its "Edit" button.

Reviewer: ```suggestion
Our `useEffect()` hook is behaving exactly as we designed it: it runs as soon as the component renders, sees that `isEditing` is `false`, and focuses the "Edit" button. There are three instances of `<Todo />`, and focus is given to the "Edit" button of the one that renders last.
```

Reviewer: Just because "renders moves focus" is a bit of a mouthful. I found myself tripping over it when reading.

}
```

> `usePrevious()` is a _custom hook_ that tracks a value across renders. It uses the `useRef()` hook to create a ref, and then updates the `useEffect()` hook. This means the `current` property of the ref will always contain the previous value of the `value` argument. You might think of it as intentionally being one value "behind" – hence the name `usePrevious()`.

Reviewer: ```suggestion
`usePrevious()` is a _custom hook_ that tracks a value across renders. It uses the `useRef()` hook to create a ref, and then uses the `useEffect()` hook to store the value of `value` in the ref, as it is after the calling component has rendered. This means the `current` property of the ref will always contain the previous value of the `value` argument. You might think of it as intentionally being one value "behind" – hence the name `usePrevious()`.
```

Reviewer: Does this work? I thought this could use a little more clarity.

Author: What about:

> `usePrevious()` is a _custom hook_ that tracks a value across renders. Its code can be tricky to reason about, so here's a breakdown of how it gets executed:
>
> 1. It uses the `useRef()` hook to create an empty ref
> 2. It returns the ref's `current` value to the component that called it
> 3. _When the component renders again_, it calls `useEffect()` and  updates the value stored in `ref.current`
> 4. Because ref.current is only updated in `useEffect()`, it's always one step behind whatever `value` is in the component's main render cycle – hence the name `usePrevious()`

Feel free to adjust this if you broadly like it. I'm on my phone so I'm just trying to throw down something serviceable :p

### Using `usePrevious()`

> Now we can define a `wasEditing` constant beneath the hooks at the top of `Todo()`. We want this constant to track the previous value of `isEditing`, so we call `usePrevious` with `isEditing` as an argument:

Reviewer: ```suggestion
Now we can define a `wasEditing` constant to track the previous value of `isEditing`; this is achieved by calling `usePrevious` with `isEditing` as an argument. Add the following inside `Todo()`, below the `useRef` lines:
```

Here we are invoking `usePrevious()` to track the length of the tasks state, like so:

> **Note:** Since we're now utilizing `usePrevious()` in two files, it might be more efficient to move the `usePrevious()` function into its own file, export it from that file, and import it where you need it. Try doing this as an exercise once you've got to the end.

> ### Using `useEffect()` to control our heading focus

Author: Yeah, fair call; I could not tell you _why_ I wrote it the way I did before. Changed per your suggestion.
